### PR TITLE
bank: session cleaner

### LINKED ISSUE #117: bank: cleaning inactive sessions
Old and Inactive sessions should be deleted from the database.

### PR DESCRIPTION
**Pull request by martinmilev**
Added Session cleaner that removes old sessions from the DB.
Fixes #117.

### CODE REVIEW

**Comment by freekman** [comment]
I am seeing code from old PR's rebase

**Comment by martinmilev** [comment]
PR is ready for review.

File: src/main/java/com/clouway/App.java [hunk 1]
```diff
@@ -9,6 +14,15 @@
 public class App {
   public static void main(String[] args) {
 
+    Thread cleaner = new Thread(
+            new SessionsCleaner(
+                    5,
```
Comment (freekman) [inline src/main/java/com/clouway/App.java]: dont like the use of magic numbers

Comment (martinmilev) [inline src/main/java/com/clouway/App.java]: how to save it? as field or a class?

File: src/main/java/com/clouway/core/SessionsCleaner.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package com.clouway.core;
+
+import org.apache.commons.lang3.time.DateUtils;
+
+import java.util.Date;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class SessionsCleaner implements Runnable {
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: missing TESTS . Why?

File: src/main/java/com/clouway/core/SessionsRepository.java [hunk 3]
```diff
@@ -29,4 +30,8 @@
    * @return the count of the sessions in Integer
    */
   Integer countSessions();
+
+  List<Session> getAll();
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsRepository.java]: why the upper methods have documentation and these don't?

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java [hunk 4]
```diff
@@ -47,4 +49,36 @@ public void countingSessions() throws Exception {
...
+  @Test
+  public void getAllSessions() throws Exception {
+    DateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm");
+    Date date = format.parse("2000-01-01 00:00:00");
+
+    repo.save(new Session("id1", "userA", date));
+    repo.save(new Session("id2", "userB", date));
+
+    List<Session> actual = repo.getAll();
+    List<Session> expected = ImmutableList.of(
```
Comment (freekman) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: what made you use ImmutableList

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java [hunk 5]
```diff
@@ -47,4 +49,36 @@ public void countingSessions() throws Exception {
...
+
+  @Test
+  public void deleteSession() throws Exception {
+    Date date = new Date();
+    repo.save(new Session("id1", "userA", date));
+    repo.save(new Session("id2", "userB", date));
+
+    repo.deleteByID("id1");
+
+    List<Session> actual = repo.getAll();
```
Comment (martinmilev) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: To get all the sessions in the db. I am asserting that after deletion there is only one session left.

File: src/main/java/com/clouway/core/SessionsCleaner.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: Could not test it. Still tying.

File: src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java [hunk 6]
```diff
@@ -47,4 +49,36 @@ public void countingSessions() throws Exception {
...
+    repo.save(new Session("id1", "userA", date));
+    repo.save(new Session("id2", "userB", date));
+
+    List<Session> actual = repo.getAll();
+    List<Session> expected = ImmutableList.of(
+            new Session("id1", "userA", date),
+            new Session("id2", "userB", date)
+    );
+
+    assertThat(actual.size(), is(2));
```
Comment (martinmilev) [inline src/test/java/com/clouway/persistent/adapter/jdbc/PersistentSessionRepositoryTest.java]: fixed.

File: src/main/java/com/clouway/core/SessionsRepository.java (hunk 3 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsRepository.java]: fixed.

File: src/main/java/com/clouway/App.java (hunk 1 above)
Comment (freekman) [inline src/main/java/com/clouway/App.java]: one solution is to wrap the integer in a object

File: src/main/java/com/clouway/App.java [hunk 7]
```diff
@@ -1,13 +1,28 @@
...
+import com.clouway.persistent.adapter.jdbc.PersistentSessionRepository;
+import com.clouway.persistent.datastore.DataStore;
 import com.google.common.base.Strings;
 
 /**
  * @author Martin Milev <martinmariusmilev@gmail.com>
  */
 public class App {
   public static void main(String[] args) {
+    final Integer sessionLife = 5;
```
Comment (freekman) [inline src/main/java/com/clouway/App.java]: try to make it as a const

Comment (martinmilev) [inline src/main/java/com/clouway/App.java]: to remove it from the constructor?

File: src/main/java/com/clouway/core/SessionsCleaner.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  private final Integer sessionLife;
+  private final SessionsRepository sessionsRepo;
+  private final MyClock clock;
+  private final Display display;
```
Comment (freekman) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: i see no purpose of this class you have added it  only to be able to test the class , do you think that using a work-around is the best prosach

File: src/test/java/com/clouway/http/servlets/HomePageServletTest.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public class HomePageServletTest {
```
Comment (freekman) [inline src/test/java/com/clouway/http/servlets/HomePageServletTest.java]: Why do i have tests for the ```HomePageServlet```

Comment (martinmilev) [inline src/test/java/com/clouway/http/servlets/HomePageServletTest.java]: Don't know why that file is here.

File: src/main/java/com/clouway/core/SessionsCleaner.java (hunk 8 above)
Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: Couldn't find a better way to test it. That is the shortest workaround other than Thread.sleep(); that I can think of.

Comment (freekman) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: aaa i can see did you try to implement the functionality for the SessionCleaner without any concurrency and then in the main to fire a tread that can call the SessionCleaner ?

Comment (martinmilev) [inline src/main/java/com/clouway/core/SessionsCleaner.java]: Something like that in the main class?
```java
new Thread() {
      private SessionsCleaner cleaner = new SessionsCleaner(
              new PersistentSessionRepository(
              new DataStore(
              new ConnectionProvider())),
              new MyServerClock());
      @Override
      public void run() {
        try{
          while (true) {
            cleaner.cleanSessions();
            Thread.sleep(10000L);
          }
        } catch (InterruptedException e) {}
      }
    }.start();
```

File: src/test/java/com/clouway/http/servlets/HomePageServletTest.java (hunk 9 above)
Comment (freekman) [inline src/test/java/com/clouway/http/servlets/HomePageServletTest.java]: this is not an answer

File: src/test/java/com/clouway/core/SessionsCleanerTest.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+  @Before
+  public void setUp() throws Exception {
+    repo = new PersistentSessionRepository(new DataStore(new ConnectionProvider()));
+    DatastoreCleaner datastoreCleaner = new DatastoreCleaner("sessions");
+    datastoreCleaner.perform();
+  }
+
+  @Test
+  public void happyPath() throws Exception {    DateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm");
```
Comment (martinmilev) [inline src/test/java/com/clouway/core/SessionsCleanerTest.java]: done

**Comment by freekman** [comment]
LGTM

**Pull request merged**